web/modules/custom/joinup_subscription/src/Form/SubscriptionDashboardForm.php

```diff
@@ -0,0 +1,175 @@
+ */
+class SubscriptionDashboardForm extends FormBase {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The entity type bundle info service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface
+   */
+  protected $entityTypeBundleInfo;
+
+  /**
+   * The Joinup relation manager service.
+   *
+   * @var \Drupal\joinup_core\JoinupRelationManagerInterface
+   */
+  protected $relationManager;
+
+  /**
+   * Constructs a new SubscriptionDashboardForm.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface $entityTypeBundleInfo
+   *   The entity type bundle info service.
+   * @param \Drupal\joinup_core\JoinupRelationManagerInterface $relationManager
+   *   The Joinup relation manager service.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, EntityTypeBundleInfoInterface $entityTypeBundleInfo, JoinupRelationManagerInterface $relationManager) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->entityTypeBundleInfo = $entityTypeBundleInfo;
+    $this->relationManager = $relationManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('entity_type.bundle.info'),
+      $container->get('joinup_core.relations_manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'joinup_subscription_dashboard';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state, AccountInterface $user = NULL) {
+    // When no user is passed we cannot show anything useful.
+    if (empty($user)) {
+      throw new \InvalidArgumentException('No user account supplied.');
+    }
+
+    $memberships = $this->relationManager->getUserGroupMembershipsByBundle($user, 'rdf_entity', 'collection');
+    $bundle_info = $this->entityTypeBundleInfo->getBundleInfo('node');
+
+    $form['description'] = [
+      '#type' => 'html_tag',
+      '#tag' => 'p',
+      '#value' => $this->t('Set your preferences to receive notifications on a per collection basis.'),
+    ];
+
+    $empty_message = $this->t('No collection memberships yet. Join one or more collections to subscribe to their content!');
+    $form['empty_text'] = [
+      '#theme' => 'status_messages',
+      '#message_list' => ['status' => [$empty_message]],
+      '#status_headings' => [
+        'status' => t('Status message'),
+        'error' => t('Error message'),
+        'warning' => t('Warning message'),
+      ],
+      '#access' => !(bool) count($memberships),
+    ];
+
+    $form['collections']['#tree'] = TRUE;
+
+    foreach ($memberships as $membership) {
```

Comment:
This doesn't quite follow the wireframe. There's a functionality missed: the _LOAD MORE_ trigger. What if a use is subscribed to 100 collections? Then the whole page will not be usable. Implementing the infinite scroll cannot be achieved here, the PR would grow exponentially. Please, open a follow-up in the epic to implement that.

## Technical aspects

We can use either a _LOAD MORE_ button or an infinite scroll technique. They are the same thing, the difference is that, with the later, instead of hitting the _LOAD MORE_ button, new items are added when you scroll in the browser at the end of the list.

There's the [Views Infinite Scroll](https://www.drupal.org/project/views_infinite_scroll) module which provides such a views pager out-of-the box. But that would require to implement our form as view. There are some issues with a form as a view:

* We cannot establish a relationship between the `og_membership` table and the group, because the group is an RDF entity. However, we can workaround this in many ways (views alters, creating a new view field plugin to hold the rendered RDF entity).
* I don't see for the moment how to expose the `subscription_bundles` field in views as a list of form select elements. Probably we can create a custom formatter that acts like a widget. By using bulk operations we can create a "Submit" action the submits the form.

The other way around is to keep the actual form and only show a limited number of memberships. Then an AJAX trigger will inject new elements in the form.

So, please, open a follow-up.

Reply:
Good point, opened a follow up to handle this: https://webgate.ec.europa.eu/CITnet/jira/browse/ISAICP-5009